source/pervasive/arithmetic/power.rs

```diff
@@ -0,0 +1,802 @@
+#[cfg(verus_keep_ghost)]
+use crate::arithmetic::math::{sub as sub1};
+
+/// This function performs exponentiation recursively, to compute `b`
+/// to the power of a natural number `e`
+pub open spec fn pow(b: int, e: nat) -> int
+    decreases e,
+{
+    if e == 0 {
+        1
+    } else {
+        b * pow(b, (e - 1) as nat)
+    }
+}
+
+/// Proof that the given integer `b` to the power of 0 is 1
+pub proof fn lemma_pow0(b: int)
+    ensures
+        pow(b, 0) == 1,
+{
+    reveal(pow);
+}
+
+/// Proof that any integer to the power of 0 is 1
+pub proof fn lemma_pow0_auto()
+    ensures
+        forall|b: int| #[trigger] pow(b, 0 as nat) == 1,
+{
+    reveal(pow);
+}
+
+/// Proof that the given integer `b` to the power of 1 is `b`
+pub proof fn lemma_pow1(b: int)
+    ensures
+        pow(b, 1) == b,
+{
+    calc! { (==)
+        pow(b, 1);
+        { reveal(pow); }
+        b * pow(b, 0);
+        { lemma_pow0(b); }
+        b * 1;
+        { lemma_mul_basics_auto(); }
+        b;
+    }
+}
+
+/// Proof that any integer to the power of 1 is itself
+pub proof fn lemma_pow1_auto()
+    ensures
+        forall|b: int| #[trigger] pow(b, 1) == b,
+{
+    reveal(pow);
+    assert forall|b: int| #[trigger] pow(b, 1) == b by {
+        lemma_pow1(b);
+    };
+}
+
+/// Proof that 0 to the power of the given positive integer `e` is 0
+pub proof fn lemma0_pow(e: nat)
+    requires
+        e > 0,
+    ensures
+        pow(0, e) == 0,
+    decreases e,
+{
+    reveal(pow);
+    lemma_mul_basics_auto();
+    if e != 1 {
+        lemma0_pow((e - 1) as nat);
+    }
+}
+
+/// Proof that 0 to the power of any positive integer is 0
+pub proof fn lemma0_pow_auto()
+    ensures
+        forall|e: nat|
+            e > 0 ==> #[trigger]
+            pow(0, e) == 0,
+{
+    reveal(pow);
+    assert forall|e: nat| e > 0 implies #[trigger]
+    pow(0, e) == 0 by {
+        lemma0_pow(e);
+    }
+}
+
+/// Proof that 1 to the power of the given natural number `e` is 1
+pub proof fn lemma1_pow(e: nat)
+    ensures
+        pow(1, e) == 1,
+    decreases e,
+{
+    reveal(pow);
+    lemma_mul_basics_auto();
+    if e != 0 {
+        lemma1_pow((e - 1) as nat);
+    }
+}
+
+/// Proof that 1 to the power of any natural number is 1
+pub proof fn lemma1_pow_auto()
+    ensures
+        forall|e: nat|
+            e > 0 ==> #[trigger]
+            pow(1, e) == 1,
+{
+    reveal(pow);
+    assert forall|e: nat| e > 0 implies #[trigger]
+    pow(1, e) == 1 by {
+        lemma1_pow(e);
+    }
+}
+
+/// Proof that taking the given number `x` to the power of 2 produces `x * x`
+pub proof fn lemma_square_is_pow2(x: int)
+    ensures
+        pow(x, 2) == x * x,
+{
+    // the original dafny is just a reveal clause
+    // maybe I can do it with reveal_with_fuel? but I don't know how
+    // the following doesn't work
+    // assert(pow(x, 2) == x * x) by {
+    // reveal_with_fuel(pow, 2);
+    //  };
```

Comment:
This works with `reveal_with_fuel(pow, 3)`:
https://play.verus-lang.org/?version=stable&mode=basic&edition=2021&gist=5bc7836e15e5205a1d1dc9582901a774
You need one fuel to learn that `pow(x, 2) == x * pow(x, 1)`, another fuel to learn `pow(x, 1) == x * pow(x, 0)`, and one more to learn what `pow(x, 0)` is.

Reply:
Thanks, updated!